### PR TITLE
Add Flash Attention 2 support to Musicgen and Musicgen Melody

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Supersedes #27924

The attention tests all pass but there are no integration equivalence between the original attention models and the FA ones. I don't hear any difference in quality despite not being the same song, though.

cc @sanchit-gandhi and @amyeroberts, could you review please?

### CODE REVIEW

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for adding this!

File: src/transformers/models/musicgen/configuration_musicgen.py [hunk 1]
```diff
@@ -239,3 +239,20 @@ def from_sub_models_config(
     # This is a property because you might want to change the codec model on the fly
     def sampling_rate(self):
         return self.audio_encoder.sampling_rate
+
+    @property
+    def _attn_implementation(self):
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/musicgen/configuration_musicgen.py:244]: Not if we want to keep the `setter` part!

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 2]
```diff
@@ -277,6 +281,300 @@ def test_greedy_generate_stereo_outputs(self):
...
+                    if not model.config.is_encoder_decoder
+                    else outputs.decoder_hidden_states[-1]
+                )
+                logits_fa = (
+                    outputs_fa.hidden_states[-1]
+                    if not model.config.is_encoder_decoder
+                    else outputs_fa.decoder_hidden_states[-1]
+                )
+
+                assert torch.allclose(logits_fa[1:], logits[1:], atol=4e-2, rtol=4e-2)
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply tests/models/musicgen/test_modeling_musicgen.py:352]: I've copied out the same tolerance threshold than any other models (regardless of modality) btw

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 3]
```diff
@@ -667,6 +1026,8 @@ def __init__(self, config: MusicgenDecoderConfig):
 
         self.layers = nn.ModuleList([MusicgenDecoderLayer(config) for _ in range(config.num_hidden_layers)])
         self.layer_norm = nn.LayerNorm(config.hidden_size)
+        self._use_flash_attention_2 = config._attn_implementation == "flash_attention_2"
+        self._use_sdpa = config._attn_implementation == "sdpa"
```
Comment (ArthurZucker) [inline src/transformers/models/musicgen/modeling_musicgen.py:1030]: let's only save `self._attn_implementation` please

File: src/transformers/models/musicgen_melody/modeling_musicgen_melody.py [hunk 4]
```diff
@@ -324,17 +341,348 @@ def forward(
...
+
+        # Use the `embed_dim` from the config (stored in the class) rather than `hidden_state` because `attn_output` can be
+        # partitioned across GPUs when using tensor-parallelism.
+        attn_output = attn_output.reshape(bsz, tgt_len, self.embed_dim)
+
+        attn_output = self.out_proj(attn_output)
+
+        return attn_output, None, past_key_value
+
+
```
Comment (ArthurZucker) [inline src/transformers/models/musicgen_melody/modeling_musicgen_melody.py:665]: copied from can be used here as well!

**Review by ArthurZucker — COMMENTED** [review]
LGTM! Tests are ... huge, would be nice if you can use copied from, would help the review 😅

**Review by ArthurZucker — APPROVED** [review]
Ouf! Thanks for the big PR and adding those tests!

**Pull request merged**